embedded-io/src/lib.rs

```diff
@@ -300,12 +300,23 @@ impl<E: fmt::Debug> std::error::Error for WriteFmtError<E> {}
 pub trait Read: ErrorType {
     /// Read some bytes from this source into the specified buffer, returning how many bytes were read.
     ///
-    /// If no bytes are currently available to read, this function blocks until at least one byte is available.
+    /// If no bytes are currently available to read:
+    /// - The method blocks until at least one byte becomes available;
+    /// - Once at least one (or more) bytes become available, a non-zero amount of those is copied to the
+    ///   beginning of `buf`, and the amount is returned, *without waiting or blocking any further for
+    ///   more bytes to become available*.
+    ///
+    /// If bytes are available to read:
+    /// - A non-zero amount of bytes is read to the beginning of `buf`, and the amount is returned immediately,
+    ///   *without blocking and waiting for more bytes to become available*;
+    /// - It is not guaranteed that *all* available bytes are returned, it is possible for the implementation to
```

Comment:
minor nit: this line looks like it applies only to the "if there are bytes available to read" case, while it should apply to both. Perhaps split it to a new paragraph and remove the `-`.

Reply:
Hmmm, but it _is_ applicable only to the "if there are bytes available to read" case, as it talks about - if there are bytes available to read _in the first place_ - how many of those would be returned - i.e. not necessarily all of them (because (a) the buffer might actually be shorter than what is available (b) because the implementation just decides so, for whatever other reasons).

Reply:
This applies to the first case as well, doesn't it? In "Once at least one (or more) bytes become available", one could also ask if it's guaranteed that all available bytes are returned. The answer is "it is not guaranteed", but it's not completely obvious from the description as it is written now.

But as Dirbaio wrote, it is a "minor nit". Personally, I would not expect to have such a guarantee when reading the text you proposed.

Reply:
Not sure how come the first case is unclear that only a subset of the just-read bytes are returned, given that it says (emphasis mine): 
"
- Once at least one (or more) bytes become available, **a non-zero amount of those** is copied to the beginning of `buf`, and the amount is returned...

"

"a non-zero amount of those". Do we really need to be more explicit than that?

Reply:
Ah be it. If two folks find it unclear then it is probably unclear. :(
I'll re-arrange so that the problematic text is a separate paragraph.

Reply:
@jannic I've re-arranged it now so that the problematic text is a separate paragraph. If you can take a quick look - thanks.

Reply:
> Not sure how come the first case is unclear that only a subset of the just-read bytes are returned, given that it says (emphasis mine): "
> 
>     * Once at least one (or more) bytes become available, **a non-zero amount of those** is copied to the beginning of `buf`, and the amount is returned...
> 
> 
> "
> 
> "a non-zero amount of those". Do we really need to be more explicit than that?

IMHO the only possible source of confusion was the asymmetry between the two cases. Yes, in general I agree that the wording "a non-zero amount of those" was sufficiently clear. But then the second case, in addition to that, explicitly stated that it's allowed to return less than available. For the first case, that sentence was missing. Readers may wonder why? Is there a difference in guarantees?

Again, this was a _very_ minor detail. But if we can avoid this possible confusion by just putting the last sentence in a separate paragraph, as you now did, we should do it.

Reply:
Argh, don't want to over-complicate it, but _now_ - with the problematic text being a separate paragraph - the _next_ paragraph that starts with "This waiting behavior is important for the cases where `Read`..." becomes unclear... in that readers might be confused that it only refers to the newly-born "Note..." paragraph, while it refers to _everything_ mentioned above.

Any brilliant ideas? Or is it just my mind that finds the new arrangement also problematic?